STM: State abstraction

mifrandir opened the pull request:
These changes aim to achieve the following:

1. Remove telemetry command acknowledgements entirely
2. Make it easier to modify `State` structure.
3. Make sure every piece of (pod) logic only appears exactly once in the code.
4. Improve the modelling of the states by introducing `State::enter` and `State::exit`.
5. Streamline transition behaviour by allowing exactly one transition function per state.
6. Group transition logic by grouping conditions in `transitions.(c/h)pp`.
7. Keep/make the module code readable from an outsiders perspective.

This is a bit different to the [original proposal](https://docs.google.com/document/d/16HMSW_4KsAZOLMpU7gmKqnUT9eL6zdWJUbkt-qn-B1A/edit?usp=sharing) as I've had a chat with Brano.

There are still a few things missing before we can merge `stm` into `develop` afterwards, but this code should be very close to the "final" version. Therefore, I've already added some reviewers that don't have a lot to do with STM directly.

Reviewers:
1. @brano2: We've talked about this, so this one is pretty obvious.
2. @efe-ozbatur, @Yining-Wang: Please check that the logic is correct. Also, you're going to be extending this code so make sure you're able to understand what's going on. If you're having trouble, we might need to simplify it a bit so that it is easier to understand. The macro might look a bit daunting but you're going to be the ones who benefit the most from it as it saves *a ton* of work. Trust me.
3. @pmorande27: Since this introduces a lot of changes I think it's a good idea to get QA onboard. Also, please don't criticise my monolithic commit... there wasn't a clean way to separate the changes as they all kind of happened at once.
4. @jamesridley: Motor controls had some questions regarding our code changes so I think it's a good idea if you have a look. Also, we want to make sure members of other projects are able to understand our code.
5. @robertasn: Similar to James. Also, make sure that the status messages are verbose enough and not too verbose. I'm not sure what the sweet spot here is, so I'll let you be the judge of that. I plan on extracting the logging in the future, but this might happen after we merge into `develop`.

mifrandir committed:
STM: Implemented exit and enter, unified transitions, removed telemetry acks and removed boilerplate code

robertasn commented on src/state_machine/transitions.cpp:
```diff
@@ -0,0 +1,177 @@
+
+State *checkLaunchCommand(Logger &log, Telemetry telemetry_data)
+{
+  if (!telemetry_data.calibrate_command) return NULL;
```
Shouldn't it be `launch_command`? As I can't see `launch_command` anywhere else in the code

mifrandir replied:
Good spot! Will fix.

robertasn commented:
Also, I don't think it is related to this pull request (as this issue is from long time ago), but could you move `telemetry_data.run_length` to `nav_data`? I don't think it belongs to Telemetry.

robertasn commented on src/state_machine/transitions.cpp:
```diff
@@ -0,0 +1,177 @@
+State *checkEnteredBrakingZone(Logger &log, Navigation &nav_data, Telemetry &telemetry_data)
+{
+  // TODO(Franz): Fix braking buffer.
+  if (telemetry_data.run_length > nav_data.displacement + nav_data.braking_distance) return NULL;
```
Also, I don't think it is related to this pull request (as this issue is from long time ago), but could you move `telemetry_data.run_length` to `nav_data`? I don't think it should belong to Telemetry.

mifrandir replied:
Yeah, I can do that.

mifrandir replied:
This is actually quite convenient for us, too. I thought about it yesterday but it wasn't relevant enough to investigate.

mifrandir committed:
Set initial state for static state instances.

mifrandir committed:
Removed Logger and Main from State

This allows the static state object to be initialised anywhere at any time, regardless of whether our thread has been started. It also makes more sense as it's Main's job to manage the relation to the rest of the software.

mifrandir committed:
Reverted to using State::getInstance() and making State::instance_ protected.

This should
1. avoid potential bugs in the future as we cannot accidentally modify the static instance (or delete it...), and
2. allow for non-static states in the future, as we can then just call the constructor in getInstance for one specific state.

mifrandir committed:
Added clang-format configuration.

This file is quite useful and I think there is any harm in adding this to the repository. In almost all cases it removes all of the formatting related linter errors.

mifrandir committed:
Fixed checkLaunchCommand to actually check launch_command

mifrandir committed:
Moved run_length to navigation and thus simplified the transition

mifrandir committed:
Removed direct access to Idling::instance_ in main.cpp

robertasn reviewed (approved):
Approving the changes that are related to Telemetry subsytem.

brano2 reviewed (commented):
Just having a quick first look here... It's not too important, but I'd put the `.clang-format` file in a separate PR. (You can then always merge if from `develop` back into here.) It doesn't have anything to do with the state machine changes, does it? You shouldn't piggyback things on top of unrelated PRs for many reasons (I can elaborate if you like :smile:). But especially not on top of ones that are already very big like this one.

brano2 commented on src/state_machine/state.cpp:
```diff
@@ -24,297 +24,306 @@ namespace hyped {
+  updateModuleData();
+
+  State *next;
+  next = checkEmergencyStationery(log, embrakes_data_, nav_data_, batteries_data_, telemetry_data_,
```
![image](https://blog.daraz.pk/wp-content/uploads/2019/03/25Mar19-5B.png)

I promise I'll also get to looking at the stuff you actually asked me to look at. :see_no_evil: :smile:

mifrandir replied:
That hurts... but I'll fix it anyway.

mifrandir committed:
Stationary, not stationery

mifrandir committed:
Removed .clang-format

mifrandir commented:
I've addressed the clang-format issue in #17.

If we can agree that this is the right way to format, we could also automatically format all the files before committing them. I'd ask QA to look into this, but I'm willing to help.

If it isn't the right way, then I'd suggest we try and find it and enforce that one.

robertasn commented:
https://github.com/Hyp-ed/hyped-2021/blob/7dc198fca297a73bf23c0aa7a5563d05787899d5/src/data/data.hpp#L157-L170 Here I think `kRunComplete`, `kExiting` and `kInvalid` states are never set by state machine, but other subsystems rely on these states.

mifrandir commented:
That is also a good observation. I think this PR works without fixing that, but I will keep in mind that we have to change that before we merge stm into develop.

brano2 commented on src/state_machine/state.cpp:
```diff
@@ -24,297 +24,306 @@ namespace hyped {
+                                 sensors_data_, motors_data_);
+
+  if (next) return next;
+  return NULL;
```
It's [probably better](https://www.quora.com/Whats-the-difference-between-NULL-and-nullptr-in-C++) to use `nullptr` instead of `NULL`.

mifrandir replied:
If that's how C++ people do it, so be it... even though I like my `NULL` :(

brano2 commented on src/state_machine/state.hpp:
```diff
@@ -39,108 +40,58 @@ class Main;  // Forward declaration
-  State(Logger &log, Main *state_machine);
+  State();
+  static State *getInstance();
+  static void initialise();
```
Is this used anywhere? If so, what is it for?

mifrandir replied:
It did have a purpose but that's not relevant anymore. I changed the static instance stuff quite a few times (and will even more) and this is left over. Thanks for reminding me.

brano2 commented on src/state_machine/state.hpp:
```diff
@@ -39,108 +40,58 @@ class Main;  // Forward declaration
+   public:                                                                                         \
+    S() {}                                                                                         \
+    static S *instance_;                                                                           \
+    static S *getInstance() { return S::instance_; }                                               \
```
Does it not work when you only define `getInstance()` in `State`?

mifrandir replied:
I think if I only do it in state it will always refer to `State::instance_` and not to `S::instance_`. Because the members are static you can't make them virtual, as far as I understand. So you have to know which types' member you need.

I can check that, though.

brano2 commented on src/state_machine/state.cpp:
```diff
@@ -24,297 +24,306 @@ namespace hyped {
+//  Idling State
+//--------------------------------------------------------------------------------------
+
+Idling *Idling::instance_ = new Idling();
```
I think it's generally better to avoid dynamic allocation for performance reasons. Couldn't you make `instance_` a value and then just return its address from `getInstance()`? If not, do you delete the pointers at some point?

mifrandir replied:
That should work, yeah. 

I just didn't really change it all that much from what was there before. Good idea though!

brano2 commented on src/state_machine/state.hpp:
```diff
@@ -39,108 +40,58 @@ class Main;  // Forward declaration
-  virtual void transitionCheck() = 0;
+  virtual State *checkTransition(Logger &log) = 0;
 
-  Logger &log_;
```
Why not have Logger as a member? Now you need to pass it to everything all the time as an argument.

mifrandir replied:
I just don't think it's the states' job. Everything that has got to do with I/O should be in main and the states should only *do* stuff if they're explicitly asked. It's not a good idea to start throwing unrelated things into an object just because it's more convenient.

brano2 replied:
That's fair, definitely not a good idea to chuck unrelated things into classes for no good reason. But one could also argue that it is a state's job to report on what it is doing for debugging purposes. And most classes in the code base have their own logger even though some of them use it much less than the states do so I think it would be reasonable to make an exception in this case.
It is unfortunate that logging cannot be handled completely separately without polluting actual useful code, but it seems to me that passing the logger as an argument all the time makes that problem worse rather than better. The ultimate solution might be to have some sort of globally accessible logging so you won't need to pass logger instances around at all but that's a whole different issue.

In any case, I don't mind too much, so whatever you decide to (not) do, I'm on board with it. :)

mifrandir replied:
Also, not giving each state the Logger makes it easier because we can initialise all the state instances right away and don't rely on Main to pass the right arguments.

brano2 replied:
Ah, ok, that's a good point :+1:

brano2 commented on src/state_machine/state.cpp:
```diff
@@ -24,297 +24,306 @@ namespace hyped {
 
-  if (nav_data.velocity <= 0) {
-    log_.INFO("STM", "zero velocity reached");
+  data::StateMachine sm_data = data_.getStateMachineData();
+  sm_data.current_state      = data::State::kFailureStopped;
+  data_.setStateMachineData(sm_data);
```
Maybe you could create a helper method for this. Something like `void State::updateStateMachineData(data::State new_state)`. Although it is just three lines so doesn't matter too much I guess :unamused:

It would be really cool if you could create some sort of centralised mapping between `State` subclasses, the values of the `data::State` enum, and the string representations of the states. That way, you could have a common implementation of the `enter()` and `exit()` methods in `State`.

mifrandir replied:
That sounds doable. We can just give constants to each substate. Something like `Idle::stringRepresentation` and `Idle::enumValue`. 

I'm going to try that.

brano2 replied:
Sounds good. I suppose you'll need to initialise them all in `states.cpp` but it might still end up neater... 🤔 I guess you'll find out 😄

brano2 commented on src/state_machine/transitions.hpp:
```diff
@@ -0,0 +1,118 @@
+// Telemetry Commands
+//--------------------------------------------------------------------------------------
+
+/*
+ * @brief   Returns the Accelerating state if the launch command has been received.
+ */
+State *checkLaunchCommand(Logger &log, Telemetry telemetry_data);
+
+/*
+ * @brief    Returns the Stopped state if the shutdown command has been received.
+ */
+State *checkShutdownCommand(Logger &log, Telemetry telemetry_data);
```
Is there much point in having these? They are pretty much one-liners and they are only used once each. It seems to me like an unnecessary fragmentation of the logic of the states that use them. Same goes for the navigation data events below.

mifrandir replied:
I tried to separate the high-level behaviour and the details. As you've pointed out above, changing the return type to boolean might be a good idea. Yet, since those functions do more than just checking the command (i.e. logging), I think it's a good idea to have them grouped together. It should remove some noise in `state.cpp`.

brano2 commented on src/state_machine/transitions.cpp:
```diff
@@ -0,0 +1,177 @@
+  if (motors_data.module_status != ModuleStatus::kInit) return NULL;
+
+  log.INFO("STM", "Calibrate command received and all modules initialised");
+  return Calibrating::getInstance();
```
I'd argue that it is really not clear that `checkModulesInitialised()` should return the `Calibrating` state. Likewise, I have no idea from which state it is supposed to be called.

I think the checks should only return a `bool` and let the `State`s decide where to transition to. If you added another state before the current initial one, you might want to use this function in it and even though the logic is the same, you'd have to go in and change which state is returned (and also change that in the comments). If you returned a `bool` from all the checks, you probably wouldn't need to touch `transitions.[ch]pp` at all (or only to add a whole new function).

To some extent, these concerns apply to all the check functions, but this is probably the worst offender.

mifrandir replied:
Agreed, see above.

You just have to draw the line somewhere and I went for the other approach.

brano2 replied:
Of course, if I was doing it, it's entirely possible that I would end up with the same thing and you'd tell me to change it 😄

brano2 commented on src/state_machine/state.hpp:
```diff
@@ -39,108 +40,58 @@ class Main;  // Forward declaration
-   */
-  void transitionCheck();
-};
+MAKE_STATE(Idling)          // State on startup
```
Not sure if _Idling_ is the best name. When talking about a car, it usually means running the engine while stationary but the pod's state is more akin to turning the ignition key just the first notch and only having a limited selection of the electrical systems come online. And couldn't idling also mean moving without accelerating or decelerating? That would be even more misleading. I think _Idle_ was better.

mifrandir replied:
I'm not an engineer, so I honestly don't see the difference between `Idle` and `Idling`. To be honest I didn't even realise I had changed something about that. 

That being said, it's an easy fix.

brano2 commented on src/state_machine/transitions.cpp:
```diff
@@ -0,0 +1,177 @@
+  return false;
+}
+
+State *checkEmergencyStationary(Logger &log, EmergencyBrakes embrakes_data, Navigation nav_data,
+                                Batteries batteries_data, Telemetry telemetry_data,
+                                Sensors sensors_data, Motors motors_data)
+{
+  if (!checkEmergencyGeneral(log, embrakes_data, nav_data, batteries_data, telemetry_data,
+                             sensors_data, motors_data)) {
+    return NULL;
+  }
+  return FailureStopped::getInstance();
+}
+
+State *checkEmergencyMoving(Logger &log, EmergencyBrakes embrakes_data, Navigation nav_data,
+                            Batteries batteries_data, Telemetry telemetry_data,
+                            Sensors sensors_data, Motors motors_data)
+{
+  if (!checkEmergencyGeneral(log, embrakes_data, nav_data, batteries_data, telemetry_data,
+                             sensors_data, motors_data)) {
+    return NULL;
+  }
+  return FailureBraking::getInstance();
+}
```
These don't do much, especially when you consider that the states that use them do the same if-statement all over again. Seems like an unnecessary fragmentation of the logic of those states - the code there would not be one bit more complex or longer if the states decided for themselves what the next state should be (and used `checkEmergencyGeneral()` directly).

mifrandir replied:
Those will be removed once I've changed the transition conditions to boolean.

brano2 reviewed (commented):
I went over the high level structure and made some comments.

Most importantly, I don't like that the `checkXYZ()` functions return a state pointer. I think it would be better if they returned `bool`s which seems like a much better separation of concerns between `states.[ch]pp` and `transitions.[ch]pp`. Currently, the transitions have no knowledge of the current state so it is weird that they can make a call as to what the next state should be. The current structure also makes it much more difficult to check that the state machine is implemented according to the diagram as I have to constantly jump between `states.cpp` and `transitions.cpp` to follow the transitions. If the checks returned `bool`s (and had good names), it should be easy to read just `states.cpp` and then check `transitions.cpp` separately to see that the helper functions do what they say they do. And they should just hide the most tedious and repetitive logic details.

Also, did you check if the style guide says anything about `*`/`&` going with the type vs with the name? I was under the impression that most of the code had them with the type but not sure.

mifrandir committed:
Removed unused initialise functions

mifrandir committed:
Changed NULL to nullptr

mifrandir committed:
Removed dynamic allocation for static state instances

mifrandir committed:
Changed transition helper functions to booleans

mifrandir committed:
Renamed Idling to Idle

mifrandir committed:
Made S::instance_ private

mifrandir committed:
Streamlined entry and exit behaviour

mifrandir committed:
Fixed descriptions in file headers

mifrandir committed:
Fixed typo in state.hpp

jamesridley commented on src/state_machine/state.cpp:
```diff
@@ -24,297 +25,192 @@ namespace hyped {
+  if (checkShutdownCommand(log, telemetry_data_)) { return Off::getInstance(); }
+  return nullptr;
 }
 
```
Do we ever enter kRunComplete or kExiting?

robertasn replied:
> https://github.com/Hyp-ed/hyped-2021/blob/7dc198fca297a73bf23c0aa7a5563d05787899d5/src/data/data.hpp#L157-L170
> 
> Here I think `kRunComplete`, `kExiting` and `kInvalid` states are never set by state machine, but other subsystems rely on these states.

> That is also a good observation. I think this PR works without fixing that, but I will keep in mind that we have to change that before we merge stm into develop.

jamesridley replied:
If I recall correctly, runComplete is when the pod has finished braking, but has not yet begun to crawl out of the tube using service propulsion. Exiting is when the pod is actively waiting for the SVPGO/SVPSTOP commands from telemetry. Finished is when the pod has completed the crawl out the tube.

jamesridley replied:
Oh i thought this pr was for a merge into develop. In any case, these states are important for motors, but could be simplified! Maybe a single state for post run, stopped, and a single state for post run, crawling. This would be easy to account for in the propulsion code, but im not sure about other subsystems

mifrandir replied:
That sounds like we're going to leave them in then. 

We never actually enter them at the moment since crawling is not a thing, but if you guys have implemented that part already, it will be easier in the future to make it work.
